src/server/ua_server.c

```diff
@@ -693,6 +693,238 @@ UA_Server_removeCallback(UA_Server *server, UA_UInt64 callbackId) {
+    } else {
+        trustList.specifiedLists = UA_TRUSTLISTMASKS_ISSUERCERTIFICATES | UA_TRUSTLISTMASKS_ISSUERCRLS;
+        trustList.issuerCertificates = certificates;
+        trustList.issuerCertificatesSize = certificatesSize;
+        trustList.issuerCrls = crls;
+        trustList.issuerCrlsSize = crlsSize;
+        fullMatchRequired = true;
+    }
+
+    UA_StatusCode retval = checkCrlMatching(certificates, certificatesSize, crls,
+                                            crlsSize, fullMatchRequired);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    /* When adding certificate files to the TrustList,
+     * it is not necessary to check the trust status of the existing SecureChannels. */
+    if(appendCertificates)
+        return certGroup->addToTrustList(certGroup, &trustList);
+
+    retval = certGroup->setTrustList(certGroup, &trustList);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    UA_DelayedCallback *dc = (UA_DelayedCallback*)UA_calloc(1, sizeof(UA_DelayedCallback));
+    if(!dc)
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+
+    dc->callback = secureChannel_delayedCloseTrustList;
+    dc->application = server;
+    dc->context = dc;
+
+    UA_EventLoop *el = server->config.eventLoop;
+    el->addDelayedCallback(el, dc);
+
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_StatusCode
+UA_Server_removeCertificates(UA_Server *server,
+                          const UA_NodeId certificateGroupId,
+                          UA_ByteString *certificates,
+                          size_t certificatesSize,
+                          const UA_Boolean isTrustedCertificates) {
+
+    UA_CertificateGroup *certGroup = getCertificateGroup(server, certificateGroupId);
+    if(!certGroup)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    UA_TrustListDataType oldTrustList;
+    UA_TrustListDataType_init(&oldTrustList);
+    UA_ByteString *crlsOld;
+    size_t crlsOldSize = 0;
+    UA_Boolean fullMatchRequired = false;
+
+    if(isTrustedCertificates) {
+        oldTrustList.specifiedLists = UA_TRUSTLISTMASKS_TRUSTEDCRLS;
+        certGroup->getTrustList(certGroup, &oldTrustList);
+        crlsOld = oldTrustList.trustedCrls;
+        crlsOldSize = oldTrustList.trustedCrlsSize;
+    }
+    else {
+        oldTrustList.specifiedLists = UA_TRUSTLISTMASKS_ISSUERCRLS;
+        certGroup->getTrustList(certGroup, &oldTrustList);
+        crlsOld = oldTrustList.issuerCrls;
+        crlsOldSize = oldTrustList.issuerCrlsSize;
+        fullMatchRequired = true;
+    }
+
+    /* There is a maximum of one CRL for each certificate.
+     * Therefore, the maximum number of CRLs to be removed is equal to the number of certificates to be removed. */
+    UA_ByteString *crls = (UA_ByteString*)UA_Array_new(certificatesSize, &UA_TYPES[UA_TYPES_BYTESTRING]);
+    size_t crlsSize = 0;
+
+    for(size_t i = 0; i < crlsOldSize; i++) {
+        for(size_t j = 0; j < certificatesSize; j++) {
```

Comment:
Break already when equality is achieved.
Otherwise you have an overflow access below.